core/jobs/batch_jobs/exp_migration_jobs_test.py

```diff
@@ -453,3 +466,817 @@ def test_unmigrated_exp_is_migrated(self) -> None:
+        exp_services.save_new_exploration(
+            feconf.SYSTEM_COMMITTER_ID, exploration)
+        self.assertEqual(
+            exploration.states_schema_version,
+            feconf.CURRENT_STATE_SCHEMA_VERSION)
+
+        # Start migration job on sample exploration.
+        self.assert_job_output_is([
+            job_run_result.JobRunResult(
+                stdout='',
+                stderr=(
+                    'EXP PROCESSED ERROR: "(\'exp_id0\', Exception(\'Snapshot'
+                    ' is already at latest schema version\'))": 1'
+                )
+            )
+        ])
+
+    def test_migration_audit_job_skips_deleted_explorations(self) -> None:
+        """Tests that the snapshot migration job skips deleted explorations
+        and does not attempt to migrate any of the snapshots.
+        """
+        exploration = exp_domain.Exploration.create_default_exploration(
+            self.VALID_EXP_ID, title='title', category='category')
+        init_state = exploration.states[exploration.init_state_name]
+        self.set_interaction_for_state(init_state, 'EndExploration')
+        init_state.update_interaction_default_outcome(None)
+        exp_services.save_new_exploration(
+            feconf.SYSTEM_COMMITTER_ID, exploration)
+
+        # Note: This creates a summary based on the upgraded model (which is
+        # fine). A summary is needed to delete the exploration.
+        exp_services.regenerate_exploration_and_contributors_summaries(
+            self.VALID_EXP_ID)
+
+        # Delete the exploration before migration occurs.
+        exp_services.delete_exploration(
+            feconf.SYSTEM_COMMITTER_ID, self.VALID_EXP_ID)
+
+        # Ensure the exploration is deleted.
+        with self.assertRaisesRegex(Exception, 'Entity .* not found'):
+            exp_fetchers.get_exploration_by_id(self.VALID_EXP_ID)
+
+        # Start migration job on sample exploration.
+        self.assert_job_output_is([
+            job_run_result.JobRunResult(
+                stdout='',
+                stderr=(
+                    'EXP PROCESSED ERROR: "(\'exp_id0\', '
+                    'Exception(\'Exploration does not exist.\'))": 2'
+                )
+            )
+        ])
+
+        # Ensure the exploration is still deleted.
+        with self.assertRaisesRegex(Exception, 'Entity .* not found'):
+            exp_fetchers.get_exploration_by_id(self.NEW_EXP_ID)
+
+    def test_migration_job_audit_success(self) -> None:
+        """Test that the audit job runs correctly on snapshots that use a
+        previous state schema.
+        """
+        swap_states_schema_version = self.swap(
+            feconf, 'CURRENT_STATE_SCHEMA_VERSION', 42)
+        swap_exp_schema_version = self.swap(
+            exp_domain.Exploration, 'CURRENT_EXP_SCHEMA_VERSION', 48)
+        with swap_states_schema_version, swap_exp_schema_version:
+            exploration = exp_domain.Exploration.create_default_exploration(
```

Comment:
This will create an exploration with the latest schema but an older schema version name, which will create problems for other migrations. Can we use an exp dict of the older schema? (Ditto elsewhere)

Reply:
Maybe use this https://github.com/DubeySandeep/oppia/blob/translation_restructure/core/jobs/batch_jobs/exp_migration_jobs_test.py#L61


(I'm introducing it for other tests in exp_migration_jobs_test)